File: fastApiProject/service/ragOptimizationService.py
"""
RAG优化服务
包含：问题优化、语义排序、上下文压缩、回复优化
"""
import os
import json
import httpx
from typing import List, Dict, Any, Optional

from dotenv import load_dotenv
from openai import OpenAI

# 从环境变量获取配置
load_dotenv()
API_KEY = os.getenv("API_KEY", "")
API_URL = os.getenv("API_URL", "")
MODEL = os.getenv("MODEL", "")
RERAN_MODEL = os.getenv("RERAN_MODEL", "qwen3-vl-rerank")
RESULT_MODEL = os.getenv("RESULT_MODEL", "")
EXTEND_MODEL = os.getenv("EXTEND_MODEL", "qwen3.5-flash")
# Rerank API配置
RERANK_API_URL = os.getenv("RERANK_API_URL", "")


class RAGOptimizationService:
    """RAG优化服务"""

    def __init__(self):
        self.client = None

    def _get_client(self) -> OpenAI:
        """获取OpenAI客户端"""
        if self.client is None:
            self.client = OpenAI(api_key=API_KEY, base_url=API_URL)
        return self.client

    async def optimize_query(self, query: str, history: List[dict] = None) -> dict:
        """
        根据上下文优化用户问题以提升检索效果

        Args:
            query: 用户原始问题
            history: 对话历史消息

        Returns:
            {
                "original_query": str,           # 原始问题
                "expanded_queries": str    # 1个扩写问题
            }
        """
        client = self._get_client()

        # 构建历史消息文本
        history_text = ""
        if history:
            history_text = "\n".join([
                f"{msg.get('role', 'user')}: {msg.get('content', '')}"
                for msg in history[-5:]  # 只取最近5条
            ])

        prompt = f"""你是一个查询优化专家。请将用户的问题从不同角度重新表述，生成5个语义相近但表述不同的问题版本，以便在知识库中检索到更相关的信息。

对话历史：
{history_text if history_text else '（无历史消息）'}

用户当前问题：{query}

请生成扩写版本，要求：
1. 根据上下文信息总结用户问题的核心语义
2. 可以适当补充上下文信息（结合对话历史）

请以JSON格式返回结果，格式如下：
{{
    "expanded_queries": "扩写版本"
}}

注意：只返回JSON，不要有其他内容"""

        try:
            response = client.chat.completions.create(
                model=EXTEND_MODEL,
                messages=[{"role": "user", "content": prompt}],
                temperature=0.1
            )

            result_text = response.choices[0].message.content.strip()

            # 尝试解析JSON
            # 处理可能的markdown代码块
            if "```json" in result_text:
                result_text = result_text.split("```json")[1].split("```")[0]
            elif "```" in result_text:
                result_text = result_text.split("```")[1].split("```")[0]

            result = json.loads(result_text)

            return {
                "original_query": query,
                "expanded_queries": result.get("expanded_queries", [])[:1]  # 最多5个
            }

        except Exception as e:
            return {
                "original_query": query,
                "expanded_queries": []
            }

    async def select_best_query(
            self,
            original_query: str,
            expanded_queries: List[str],
            threshold: float = 0.8
    ) -> str:
        """
        使用 Rerank 选择最佳扩写问题

        Args:
            original_query: 原始问题
            expanded_queries: 扩写的问题列表
            threshold: 相关性阈值，低于此值返回原问题

        Returns:
            最佳问题（如果都低于阈值则返回原问题）
        """
        if not expanded_queries:
            return original_query

        try:
            # 将扩写问题作为 documents 进行 rerank
            documents = [{"text": q} for q in expanded_queries]

            async with httpx.AsyncClient() as client:
                response = await client.post(
                    RERANK_API_URL,
                    headers={
                        "Authorization": f"Bearer {API_KEY}",
                        "Content-Type": "application/json"
                    },
                    json={
                        "model": RERAN_MODEL,
                        "input": {
                            "query": {"text": original_query},
                            "documents": documents
                        },
                        "parameters": {
                            "return_documents": True,
                            "top_n": 1  # 只需要得分最高的一个
                        }
                    },
                    timeout=5.0
                )

                if response.status_code != 200:
                    return original_query

                result = response.json()
                results = result.get("output", {}).get("results", [])

                if not results:
                    return original_query

                # 获取得分最高的结果
                best = results[0]
                relevance_score = best.get("relevance_score", 0)
                best_query = best.get("document", {}).get("text", "")

                # 阈值检查
                if relevance_score >= threshold and best_query:
                    return best_query

                return original_query

        except Exception:
            return original_query

    async def semantic_rerank(self, query: str, contexts: List, top_k: int = 5, threshold: float = 0.6) -> List[dict]:
        """
        对检索结果进行语义重排序（使用阿里云Rerank API）

        Args:
            query: 用户问题
            contexts: 检索到的上下文列表（支持List[dict]或List[str]）
            top_k: 返回的top k个结果
            threshold: 相关性阈值，低于此值的结果将被过滤

        Returns:
            重排序后的上下文列表
        """
        if not contexts:
            return []

        try:
            # 构建文档列表，支持dict和str两种格式
            documents = []
            for ctx in contexts:
                if isinstance(ctx, dict):
                    documents.append({"text": ctx.get('content', '')})
                elif isinstance(ctx, str):
                    documents.append({"text": ctx})
                else:
                    documents.append({"text": str(ctx.content)})
            # 调用阿里云Rerank API
            async with httpx.AsyncClient() as client:
                response = await client.post(
                    RERANK_API_URL,
                    headers={
                        "Authorization": f"Bearer {API_KEY}",
                        "Content-Type": "application/json"
                    },
                    json={
                        "model": RERAN_MODEL,
                        "input": {
                            "query": {"text": query},
                            "documents": documents
                        },
                        "parameters": {
                            "return_documents": True,
                            "top_n": min(top_k, len(contexts))
                        }
                    },
                    timeout=5.0
                )

                if response.status_code != 200:
                    # print(f"Rerank API调用失败: {response.status_code} - {response.text}")
                    return contexts[:top_k] if isinstance(contexts[0], dict) else [{"content": c, "similarity": 0} for c
                                                                                   in contexts[:top_k]]

                result = response.json()
                # print(f"Rerank API返回: {result}")

                # 解析结果
                results = result.get("output", {}).get("results", [])
                ranked_contexts = []

                for item in results:
                    idx = item.get("index", 0)
                    relevance_score = item.get("relevance_score", 0)
                    document_text = item.get("document", {}).get("text", "")

                    # 阈值筛选：只保留相关性分数高于阈值的结果
                    if relevance_score < threshold:
                        # print(f"文档{idx}相关性分数{relevance_score:.4f}低于阈值{threshold}，已过滤")
                        continue

                    if 0 <= idx < len(contexts):
                        original_ctx = contexts[idx]
                        # 构建返回结果
                        if isinstance(original_ctx, dict):
                            ctx = original_ctx.copy()
                            ctx['rerank_score'] = relevance_score
                            ctx['similarity'] = relevance_score
                        else:
                            ctx = {
                                "content": document_text or (
                                    original_ctx if isinstance(original_ctx, str) else str(original_ctx)),
                                "rerank_score": relevance_score,
                                "similarity": relevance_score
                            }
                        ranked_contexts.append(ctx)

                # print(f"Rerank完成，过滤前{len(results)}个，过滤后返回{len(ranked_contexts)}个结果")
                return ranked_contexts

        except httpx.TimeoutException:
            # print("Rerank API超时")
            return contexts[:top_k] if isinstance(contexts[0], dict) else [{"content": c, "similarity": 0} for c in
                                                                           contexts[:top_k]]
        except Exception as e:
            # print(f"语义排序失败: {str(e)}")
            return contexts[:top_k] if contexts and isinstance(contexts[0], dict) else [{"content": c, "similarity": 0}
                                                                                        for c in contexts[
                                                                                                 :top_k]] if contexts else []

    async def compress_context(
            self,
            query: str,
            contexts: List[dict],
            max_length: int = 3000
    ) -> str:
        """
        压缩上下文，保留与问题最相关的信息

        Args:
            query: 用户问题
            contexts: 上下文列表
            max_length: 压缩后的最大长度

        Returns:
            压缩后的上下文文本
        """
        if not contexts:
            return ""

        client = self._get_client()

        # 计算当前上下文总长度
        total_length = sum(len(c.get('content', '')) for c in contexts)

        if total_length <= max_length:
            # 不需要压缩
            return "\n\n".join([
                f"【参考资料{i + 1}】\n{c.get('content', '')}"
                for i, c in enumerate(contexts)
            ])

        # 构建上下文文本
        context_text = "\n\n".join([
            f"【参考资料{i + 1}】\n{c.get('content', '')}"
            for i, c in enumerate(contexts)
        ])

        prompt = f"""你是一个信息提取专家。请从以下参考资料中提取与用户问题最相关的关键信息。

用户问题：{query}

参考资料：
{context_text}

请执行以下操作：
1. 提取与问题直接相关的关键信息
2. 去除无关的描述和冗余内容
3. 保留重要的细节和数据
4. 保持信息的准确性，不要添加原文没有的信息

请输出压缩后的内容，保留资料编号以便追溯。压缩后的内容总长度不要超过{max_length}个字符。"""

        try:
            response = client.chat.completions.create(
                model=EXTEND_MODEL,
                messages=[{"role": "user", "content": prompt}],
                temperature=0.1
            )

            compressed = response.choices[0].message.content.strip()
            return compressed

        except Exception:
            # 返回截断的原始上下文
            result = ""
            for i, c in enumerate(contexts):
                content = c.get('content', '')
                if len(result) + len(content) > max_length:
                    break
                result += f"【参考资料{i + 1}】\n{content}\n\n"
            return result

    async def optimize_response(
            self,
            query: str,
            contexts: List[dict],
            response: str
    ) -> dict:
        """
        优化回复，判断合理性并去除不合理部分

        Args:
            query: 用户问题
            contexts: 使用的上下文
            response: 初步回复

        Returns:
            {
                "is_reasonable": bool,
                "optimized_response": str,
                "issues": List[str],
                "confidence": float  # 置信度 0-1
            }
        """
        client = self._get_client()

        # 构建上下文摘要
        context_summary = "\n".join([
            f"[资料{i + 1}] {c.get('content', '')[:150]}..."
            for i, c in enumerate(contexts[:5])
        ])

        prompt = f"""你是一个回答质量审核专家。请审核以下回答的质量并进行优化。

用户问题：{query}

参考的知识库内容：
{context_summary}

AI的回答：
{response}

请执行以下任务：
1. 判断回答是否合理（是否回答了问题、是否有依据、是否有幻觉）
2. 找出回答中的问题（如：与参考资料不符的内容、过度推断、无关信息等）
3. 优化回答，去除不合理部分

请以JSON格式返回结果：
{{
    "is_reasonable": true/false,
    "issues": ["问题1", "问题2"],
    "optimized_response": "优化后的回答",
    "confidence": 0.85
}}

注意：
- 如果回答整体合理，is_reasonable为true，issues为空列表，optimized_response可以与原回答相同
- confidence表示对回答质量的信心程度，0-1之间
- 只返回JSON，不要有其他内容"""

        try:
            llm_response = client.chat.completions.create(
                model=RESULT_MODEL,
                messages=[{"role": "user", "content": prompt}],
                temperature=0.1
            )

            result_text = llm_response.choices[0].message.content.strip()

            # 处理可能的markdown代码块
            if "```json" in result_text:
                result_text = result_text.split("```json")[1].split("```")[0]
            elif "```" in result_text:
                result_text = result_text.split("```")[1].split("```")[0]

            result = json.loads(result_text)

            return {
                "is_reasonable": result.get("is_reasonable", True),
                "optimized_response": result.get("optimized_response", response),
                "issues": result.get("issues", []),
                "confidence": result.get("confidence", 0.5)
            }

        except Exception:
            return {
                "is_reasonable": True,
                "optimized_response": response,
                "issues": [],
                "confidence": 0.5
            }


# 全局实例
rag_optimization_service = RAGOptimizationService()

File: blog-agent/agent/src/main/java/com/xiafan/agent/service/RagOptimizationService.java
package com.xiafan.agent.service;

import com.fasterxml.jackson.databind.JsonNode;
import com.xiafan.agent.config.AppProperties;
import com.xiafan.agent.service.llm.OpenAiClient;
import io.agentscope.core.message.ContentBlock;
import io.agentscope.core.message.Msg;
import io.agentscope.core.message.TextBlock;
import io.agentscope.core.message.UserMessage;
import io.agentscope.core.model.ChatResponse;
import io.agentscope.core.model.GenerateOptions;
import io.agentscope.core.model.Model;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/** Mirrors ragOptimizationService.py (query optimization, rerank, context compression, response optimization). */
@Service
public class RagOptimizationService {

    private static final Logger log = LoggerFactory.getLogger(RagOptimizationService.class);

    private final OpenAiClient openAi;
    private final AppProperties props;
    private final Model model;

    public RagOptimizationService(OpenAiClient openAi, Model model, AppProperties props) {
        this.openAi = openAi;
        this.model = model;
        this.props = props;
    }

    public Map<String, Object> optimizeQuery(String query, List<Map<String, Object>> history) {
        StringBuilder historyText = new StringBuilder();
        if (history != null && !history.isEmpty()) {
            List<Map<String, Object>> recent = history.size() > 5 ? history.subList(history.size() - 5, history.size()) : history;
            for (Map<String, Object> msg : recent) {
                if (!historyText.isEmpty()) {
                    historyText.append("\n");
                }
                historyText.append(msg.getOrDefault("role", "user")).append(": ").append(msg.getOrDefault("content", ""));
            }
        }
        String prompt = """
                你是一个查询优化专家。请将用户的问题从不同角度重新表述，生成5个语义相近但表述不同的问题版本，以便在知识库中检索到更相关的信息。

                对话历史：
                %s

                用户当前问题：%s

                请生成扩写版本，要求：
                1. 根据上下文信息总结用户问题的核心语义
                2. 可以适当补充上下文信息（结合对话历史）

                请以JSON格式返回结果，格式如下：
                {
                    "expanded_queries": "扩写版本"
                }

                注意：只返回JSON，不要有其他内容""".formatted(
                historyText.isEmpty() ? "（无历史消息）" : historyText, query);

        Map<String, Object> result = new HashMap<>();
        result.put("original_query", query);
        result.put("expanded_queries", List.of());
        try {
            String content = completeText(props.getApi().getExtendModel(), prompt, 0.1);
            if (content != null) {
                String parsed = stripMarkdownFence(content.strip());
                JsonNode json = openAi.parseTree(parsed);
                List<String> expanded = readStrings(json.get("expanded_queries"));
                result.put("expanded_queries", expanded.isEmpty() ? List.of() : List.of(expanded.get(0)));
            }
        } catch (Exception e) {
            log.warn("optimize_query failed, falling back to original query: {}", e.getMessage());
        }
        return result;
    }

    /** Re-ranks dict contexts; on failure returns the original contexts (top_k truncated). */
    public List<Map<String, Object>> semanticRerank(String query, List<Map<String, Object>> contexts, int topK, double threshold) {
        if (contexts == null || contexts.isEmpty()) {
            return List.of();
        }
        try {
            int n = Math.min(topK, contexts.size());
            Map<String, Object> body = buildRerankBody(query, contexts, n);
            JsonNode resp = openAi.parseTree(openAi.postAndGet(props.getRerank().getApiUrl(), body, 5).body());
            JsonNode results = resp.path("output").path("results");
            if (!results.isArray()) {
                return contexts.subList(0, n);
            }
            List<Map<String, Object>> ranked = new ArrayList<>();
            for (JsonNode item : results) {
                int idx = item.path("index").asInt(0);
                double score = item.path("relevance_score").asDouble(0.0);
                if (score < threshold) {
                    continue;
                }
                if (idx >= 0 && idx < contexts.size()) {
                    Map<String, Object> ctx = new LinkedHashMap<>(contexts.get(idx));
                    ctx.put("rerank_score", score);
                    ctx.put("similarity", score);
                    ranked.add(ctx);
                }
            }
            return ranked;
        } catch (Exception e) {
            log.warn("semantic_rerank failed, using original order: {}", e.getMessage());
            return contexts.subList(0, Math.min(topK, contexts.size()));
        }
    }

    /** Returns context text, compressing via LLM only when over maxLength. */
    public String compressContext(String query, List<Map<String, Object>> contexts, int maxLength) {
        if (contexts == null || contexts.isEmpty()) {
            return "";
        }
        String joined = joinContexts(contexts);
        int totalLength = 0;
        for (Map<String, Object> c : contexts) {
            Object content = c.get("content");
            if (content != null) {
                totalLength += content.toString().length();
            }
        }
        if (totalLength <= maxLength) {
            return joined;
        }
        String prompt = """
                你是一个信息提取专家。请从以下参考资料中提取与用户问题最相关的关键信息。

                用户问题：%s

                参考资料：
                %s

                请执行以下操作：
                1. 提取与问题直接相关的关键信息
                2. 去除无关的描述和冗余内容
                3. 保留重要的细节和数据
                4. 保持信息的准确性，不要添加原文没有的信息

                请输出压缩后的内容，保留资料编号以便追溯。压缩后的内容总长度不要超过%d个字符。"""
                .formatted(query, joined, maxLength);
        try {
            String compressed = completeText(props.getApi().getExtendModel(), prompt, 0.1);
            if (compressed == null || compressed.isBlank()) {
                throw new RuntimeException("empty compressed result");
            }
            return compressed.strip();
        } catch (Exception e) {
            log.warn("compress_context failed, truncating: {}", e.getMessage());
            StringBuilder result = new StringBuilder();
            for (int i = 0; i < contexts.size(); i++) {
                Object content = contexts.get(i).get("content");
                String text = content == null ? "" : content.toString();
                if (result.length() + text.length() > maxLength) {
                    break;
                }
                result.append("【参考资料").append(i + 1).append("】\n").append(text).append("\n\n");
            }
            return result.toString();
        }
    }

    public Map<String, Object> optimizeResponse(String query, List<Map<String, Object>> contexts, String response) {
        StringBuilder contextSummary = new StringBuilder();
        int limit = Math.min(5, contexts == null ? 0 : contexts.size());
        for (int i = 0; i < limit; i++) {
            Object content = contexts.get(i).get("content");
            String text = content == null ? "" : content.toString();
            String truncated = text.length() > 150 ? text.substring(0, 150) : text;
            contextSummary.append("[资料").append(i + 1).append("] ").append(truncated).append("...\n");
        }
        String prompt = """
                你是一个回答质量审核专家。请审核以下回答的质量并进行优化。

                用户问题：%s

                参考的知识库内容：
                %s

                AI的回答：
                %s

                请执行以下任务：
                1. 判断回答是否合理（是否回答了问题、是否有依据、是否有幻觉）
                2. 找出回答中的问题（如：与参考资料不符的内容、过度推断、无关信息等）
                3. 优化回答，去除不合理部分

                请以JSON格式返回结果：
                {
                    "is_reasonable": true/false,
                    "issues": ["问题1", "问题2"],
                    "optimized_response": "优化后的回答",
                    "confidence": 0.85
                }

                注意：
                - 如果回答整体合理，is_reasonable为true，issues为空列表，optimized_response可以与原回答相同
                - confidence表示对回答质量的信心程度，0-1之间
                - 只返回JSON，不要有其他内容""".formatted(query, contextSummary, response);

        Map<String, Object> result = new HashMap<>();
        result.put("is_reasonable", true);
        result.put("optimized_response", response);
        result.put("issues", List.of());
        result.put("confidence", 0.5);
        try {
            String content = completeText(props.getApi().getResultModel(), prompt, 0.1);
            if (content != null) {
                JsonNode json = openAi.parseTree(stripMarkdownFence(content.strip()));
                result.put("is_reasonable", json.path("is_reasonable").asBoolean(true));
                result.put("optimized_response", json.path("optimized_response").asText(response));
                List<String> issues = new ArrayList<>();
                JsonNode issueArr = json.path("issues");
                if (issueArr.isArray()) {
                    issueArr.forEach(i -> issues.add(i.asText()));
                }
                result.put("issues", issues);
                result.put("confidence", json.path("confidence").asDouble(0.5));
            }
        } catch (Exception e) {
            log.warn("optimize_response failed, keeping original response: {}", e.getMessage());
        }
        return result;
    }

    private String completeText(String modelName, String prompt, double temperature) {
        StringBuilder result = new StringBuilder();
        GenerateOptions options = GenerateOptions.builder()
                .modelName(modelName)
                .temperature(temperature)
                .stream(false)
                .build();
        model.stream(List.<Msg>of(new UserMessage(prompt)), List.of(), options)
                .toStream()
                .forEach(response -> result.append(responseText(response)));
        return result.toString();
    }

    private static String responseText(ChatResponse response) {
        if (response == null) {
            return "";
        }
        StringBuilder content = new StringBuilder();
        for (ContentBlock block : response.getContent()) {
            if (block instanceof TextBlock textBlock) {
                content.append(textBlock.getText());
            }
        }
        return content.toString();
    }

    private String joinContexts(List<Map<String, Object>> contexts) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < contexts.size(); i++) {
            if (i > 0) {
                sb.append("\n\n");
            }
            Object content = contexts.get(i).get("content");
            sb.append("【参考资料").append(i + 1).append("】\n").append(content == null ? "" : content);
        }
        return sb.toString();
    }

    private Map<String, Object> buildRerankBody(String query, List<Map<String, Object>> contexts, int topK) {
        Map<String, Object> body = new HashMap<>();
        body.put("model", props.getRerank().getModel());
        Map<String, Object> input = new HashMap<>();
        input.put("query", Map.of("text", query));
        List<Map<String, Object>> docs = new ArrayList<>();
        for (Map<String, Object> ctx : contexts) {
            docs.add(Map.of("text", String.valueOf(ctx.get("content"))));
        }
        input.put("documents", docs);
        body.put("input", input);
        body.put("parameters", Map.of("return_documents", true, "top_n", topK));
        return body;
    }

    private List<String> readStrings(JsonNode node) {
        List<String> out = new ArrayList<>();
        if (node == null || node.isMissingNode() || node.isNull()) {
            return out;
        }
        if (node.isArray()) {
            node.forEach(n -> out.add(n.asText()));
        } else if (node.isTextual()) {
            out.add(node.asText());
        }
        return out;
    }

    private static String stripMarkdownFence(String text) {
        if (text.contains("```json")) {
            return text.split("```json")[1].split("```")[0].trim();
        } else if (text.contains("```")) {
            return text.split("```")[1].split("```")[0].trim();
        }
        return text;
    }
}
